proxy/src/control/cache.rs

```diff
@@ -37,54 +46,85 @@ impl<T> Exists<T> {
 
 // ===== impl Cache =====
 
-impl<T> Cache<T> where T: Clone + Copy + Eq + std::hash::Hash {
+impl<K, V> Cache<K, V>
+where
+    K: Copy + Clone,
+    K: Hash + Eq,
+    V: PartialEq + Clone,
+{
     pub fn new() -> Self {
         Cache {
-            values: HashSet::new(),
+            values: IndexMap::new(),
             reset_on_next_modification: true,
         }
     }
 
-    pub fn values(&self) -> &HashSet<T> {
-        &self.values
-    }
-
     pub fn set_reset_on_next_modification(&mut self) {
         self.reset_on_next_modification = true;
     }
 
-    pub fn extend<I, F>(&mut self, iter: I, on_change: &mut F)
-        where I: Iterator<Item = T>,
-              F: FnMut(T, CacheChange),
+    /// Update the cache to contain the union of its current contents and the
+    /// key-value pairs in `iter`. Pairs not present in the cache will be
+    /// inserted, and keys present in both the cache and the iterator will be
+    /// updated so that their values match those in the iterator.
+    pub fn update_union<I, F>(&mut self, iter: I, on_change: &mut F)
+    where
+        I: Iterator<Item = (K, V)>,
+        F: FnMut((K, V), CacheChange),
     {
-        fn extend_inner<T, I, F>(values: &mut HashSet<T>, iter: I, on_change: &mut F)
-            where T: Copy + Eq + std::hash::Hash, I: Iterator<Item = T>, F: FnMut(T, CacheChange)
+        fn update_inner<K, V, I, F>(
+            values: &mut IndexMap<K, V>,
+            iter: I,
+            on_change: &mut F
+        )
+        where
+            K: Eq + Hash + Copy + Clone,
+            V: PartialEq + Clone,
+            I: Iterator<Item = (K, V)>,
+            F: FnMut((K, V), CacheChange),
         {
-            for value in iter {
-                if values.insert(value) {
-                    on_change(value, CacheChange::Insertion);
+            for (key, value) in iter {
+                match values.insert(key, value.clone()) {
+                    // If the returned value is equal to the inserted value,
+                    // then the inserted key was already present in the map
+                    // and the value is unchanged. Do nothing.
+                    Some(ref old_value) if old_value == &value => {},
```

Comment:
NIT: Elsewhere in the codebase I think we use `()` instead of `{}`. At least, that's what I've been cargo-culting.

NIT: IMO the comments are too verbose and should be removed since they're just restating the code.